### PR TITLE
System design documentation

### PR DESCRIPTION
**Pull request by mitruly**
Document VinylDNS system design information:
- Overview
- Architecture
- Components
- Integration

### CODE REVIEW

File: SYSTEM_DESIGN.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+* DynamoDB - key-value data store currently used for zone change, record set, record set change, user, group, group change and membership data.
+
+* RDS MySQL - relational database currently used for batch change and zone data.
+
+## Integration
+
+Integrating with VinylDNS is simple since each API endpoint is effectively a distinct DNS operation (eg. create record, update record, delete record, etc.). The only requirement for sending a request is generating the correct AWS SIG4 signature without content length and providing the corresponding HTTP headers so that VinylDNS can verify it.
+
+Current tooling that is available to perform VinylDNS API requests include:
```
Comment (rebstar6) [inline SYSTEM_DESIGN.md]: do we want to include tools in the diagram? it right now indicates that all requests go through the portal, which isnt the case. requests can go through the portal (to the api) OR via a tool that would hit the api directly

File: SYSTEM_DESIGN.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+* HAProxy - load balancing to manage and direct traffic. Both the portal and API service are deployed behind GSLB-VIP-HAProxy layers. Users are directed to the active instances that are geographically closer to them.
+
+* Portal - graphical user interface to interact with the VinylDNS API.
+
+* API - endpoints to interact with the VinylDNS.
+
+* SQS - message queue used to store DNS requests, allowing decoupling of processing for concurrent, highly available processing. 
+
+* DynamoDB - key-value data store currently used for zone change, record set, record set change, user, group, group change and membership data.
```
Comment (rebstar6) [inline SYSTEM_DESIGN.md]: given our plans to make the repos pluggable, i wonder if it makes sense to go into so much detail in *our* implementation. all that matters to the world should be that there needs to be one database for each of the tables we have. they can store the data wherever

File: SYSTEM_DESIGN.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+To optimize the user request processing, tasks are split into two parts: data validation and processing. Data validation includes validating well-formed user requests, valid data values and context (eg. proper authorization). Processing includes generating the DNS message and communicating with the DNS backend to apply the changes.
+
+The current implementation of VinylDNS supports performance of heavy processing tasks as follows:
+
+| Task          | Record Count |   Time  |
+|---------------|-------------:|--------:|
+| Batch change  |           20 |  ~2 min |
+| Zone sync     |        1,000 |  ~1 sec |
+| Zone sync     |       10,000 | ~10 sec |
+| Zone sync     |      100,000 |  ~6 min |
```
Comment (rebstar6) [inline SYSTEM_DESIGN.md]: similar to the comment below, these numbers are totally based on our specific implementation (including the read/write capacity on our tables i imagine). I dont know that it has much meaning to the rest of the community.

File: SYSTEM_DESIGN.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+When the user interacts with the portal, requests are made to the VinylDNS API by invoking the appropriate service calls. The code for these requests is housed in `modules/portal/public/lib`.
+
+
+## Components
+
+The following diagram illustrates the major components in the VinylDNS ecosystem and the external systems they interact with each other.
+
+![VinylDNS Architecture Diagram](vinyldns_architecture_diagram.png) 
+
+* HAProxy - load balancing to manage and direct traffic. Both the portal and API service are deployed behind GSLB-VIP-HAProxy layers. Users are directed to the active instances that are geographically closer to them.
```
Comment (pauljamescleary) [inline SYSTEM_DESIGN.md]: For this Architecture bit, we need to be a little more generic.  The only things I think we would capture are:

1. Portal
1. LDAP
1. API
1. Message Queue
1. Database

We should have those in the diagram.  Things like HAProxy are a design decision we made.

We would then mention that for the `Database`, we are using RDS + DynamoDB.

We would also mention in the detail that for the `Message Queue`, we are using `AWS SQS`

File: SYSTEM_DESIGN.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+## Table of Contents
+- [Overview](#overview)
+- [Components](#components)
+- [Process Flow](#process-flow)
+- [Integration](#integration)
+
+## Overview
+
+VinylDNS is a DNS management platform that supports user self-service resource record management by zone, currently packaged with DNS record validations, fine-grained access control, historical data tracking and support for batch DNS requests.
```
Comment (pauljamescleary) [inline SYSTEM_DESIGN.md]: I think we can ditch this overview.  Just go right to components

File: SYSTEM_DESIGN.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+* LDAP - application protocol used to authenticate user access to the VinylDNS portal.
+
+* Portal - graphical user interface to interact with the VinylDNS API.
+
+* Tooling - external libraries and utilities used to interact with the VinylDNS API.
+
+## Process Flow
+
+The portal consists of a javascript-based user interface and a thin server-side service. The portal server mostly makes API calls using the signed in user's credentials to the VinylDNS API.
```
Comment (pauljamescleary) [inline SYSTEM_DESIGN.md]: process flow should use "numbers" for steps

File: modules/api/src/main/scala/vinyldns/api/repository/mysql/VinylDNSJDBC.scala [hunk 7]
```diff
@@ -69,7 +69,7 @@ class VinylDNSJDBC(config: Config) {
   val migration = new Flyway()
   migration.setDataSource(migrationDataSource)
   // flyway changed the default schema table name in v5.0.0; this allows to revert to an old naming convention if needed
-  if(config.hasPath("default.migrationSchemaTable")) {
+  if (config.hasPath("default.migrationSchemaTable")) {
```
Comment (nimaeskandary) [inline modules/api/src/main/scala/vinyldns/api/repository/mysql/VinylDNSJDBC.scala]: +1, my editor has been adding this change constantly

File: SYSTEM_DESIGN.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+* API - RESTful endpoints to allow interaction with VinylDNS.
+
+* Database - used to store information that the VinylDNS application needs.
+
+* DNS servers - used for communicating DNS changes and resolving DNS records.
+
+* Message queue - used to store DNS requests, allowing decoupling of processing for concurrent, highly available processing. 
+
+* LDAP - application protocol used to authenticate user access to the VinylDNS portal.
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: LDAP service. based on the diagram.

File: SYSTEM_DESIGN.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+- [Process Flow](#process-flow)
+- [Integration](#integration)
+
+## Components
+
+The following diagram illustrates the major components in the VinylDNS ecosystem and the external systems they interact with each other.
+
+![VinylDNS Architecture Diagram](img/VinylDNS_overview.png)
+
+* API - RESTful endpoints to allow interaction with VinylDNS.
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: do you need the "to" in this sentence?

File: SYSTEM_DESIGN.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+* Database - used to store information that the VinylDNS application needs.
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: can we make these sentences active rather than passive? "stores information that the VinylDNS application needs."

File: SYSTEM_DESIGN.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+* LDAP - application protocol used to authenticate user access to the VinylDNS portal.
+
+* Portal - graphical user interface to interact with the VinylDNS API.
+
+* Tooling - external libraries and utilities used to interact with the VinylDNS API.
+
+## Process Flow
+
+The portal consists of a Javascript-based user interface and a thin server-side service. The portal server mostly makes API calls using the signed in user's credentials to the VinylDNS API. An alternative to the portal to interact with the VinylDNS API is to use an [integration tool](#integration) to send signed requests to the VinylDNS API.
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: do you need to say "integration tool", can it just be "integration"?

File: SYSTEM_DESIGN.md (hunk 9 above)
Comment (mitruly) [reply SYSTEM_DESIGN.md]: Yes. This describes what API is, rather than what it does.

File: SYSTEM_DESIGN.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+                                                                                                      
+1. An incoming JSON user request is received by the appropriate API endpoint (eg. batch change, membership, record set, zone, etc.), which performs JSON validations to ensure that the request is well-formed. Upon failure, an appropriate response is sent from the API endpoint back to the user. On successful JSON validation, the request is forwarded to the appropriate service for additional validations.
+
+1. The service (eg. batch change, record set, zone, membership, etc.) is responsible for gathering context for the request (eg. current record state, zone information, user authorization, etc.) and performing both data-dependent and data-independent validations. Upon failure, the response will be sent back to the user. On success, the request is serialized into a message and queued on a message queue for processing. The validation response is stored in the database.
```
Comment (pauljamescleary) [inline SYSTEM_DESIGN.md]: I think we are conflating a number of different types of services (batch change vs. record set updates) here.  To avoid that, we should be at a higher level.  The flow of processing is more like

1. When the API receives a request, it loads the credentials for the calling user from the database and validates the request signature to ensure that the request was not modified in transit
1. The request is then validated to ensure that a) the request data is correct b) the request can be passes all validation checks and c) the user has access to make the change
2. Assuming the request is in good order, the request is put on a message queue for handling
3. Background flow to continue

I think that is all that is needed.

File: SYSTEM_DESIGN.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+The process flow for the engine, which is responsible for processing the DNS request, is as follows:
+
+1. A background process continually polls the message queue. Any messages that are found are converted back to a DNS request and sent to the appropriate handler (eg. record set, zone change) to process the request. Each change is applied in an idempotent fashion, the result of applying the change is stored in the database and the message is removed from the queue.
+
+## Integration
+
+Integrating with VinylDNS is simple since each API endpoint is effectively a distinct DNS operation (eg. create record, update record, delete record, etc.). The only requirement for sending a request is generating the correct AWS SIG4 signature without content length and providing the corresponding HTTP headers so that VinylDNS can verify it.
```
Comment (pauljamescleary) [inline SYSTEM_DESIGN.md]: link to SIG4 docs (we have a section on this already)

**Comment by pauljamescleary** [comment]
I also think that the process flow should "follow the arrows" in the diagram, so that each component is covered.  I think I only had 3 "steps" in my comment, but there are clearly more arrows and boxes.

File: SYSTEM_DESIGN.md [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+The following diagram illustrates the major components in the VinylDNS ecosystem and the external systems they interact with each other.
+
+![VinylDNS Architecture Diagram](img/VinylDNS_overview.png)
+
+* API - RESTful endpoints to allow interaction with VinylDNS.
+
+* Database - stores information that the VinylDNS application needs.
+
+* DNS servers - used for communicating DNS changes and resolving DNS records.
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: this is still passive. "Communicates DNS changes and resolves DNS records."

File: SYSTEM_DESIGN.md [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+* Message queue - used to store DNS requests, allowing decoupling of processing for concurrent, highly available processing.
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: also passive. and seems wordy. can this be reworked into two sentences.

File: SYSTEM_DESIGN.md [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+## Components
+
+The following diagram illustrates the major components in the VinylDNS ecosystem and the external systems they interact with each other.
+
+![VinylDNS Architecture Diagram](img/VinylDNS_overview.png)
+
+* API - RESTful endpoints to allow interaction with VinylDNS.
+
+* Database - stores information that the VinylDNS application needs.
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: should we say what this information is, or mention that we use different types of databases?

Comment (mitruly) [reply SYSTEM_DESIGN.md]: @britneywright I had this information detailed before, but it was discussed in earlier review comments that component information should be generic and high-level.

File: SYSTEM_DESIGN.md (hunk 15 above)
Comment (mitruly) [reply SYSTEM_DESIGN.md]: There's not really a reason to have two sentences for this, since they are related and not separate ideas. If anything, I'd just shorten this to "stores DNS requests".

File: SYSTEM_DESIGN.md [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+- [Process Flow](#process-flow)
+- [Integration](#integration)
+
+## Components
+
+The following diagram illustrates the major components in the VinylDNS ecosystem and the external systems they interact with each other.
+
+![VinylDNS Architecture Diagram](img/VinylDNS_overview.png)
+
+* API - RESTful endpoints to allow interaction with VinylDNS.
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: remove the period here or add periods to the ends of the other points.

File: SYSTEM_DESIGN.md [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+1. User navigates portal or uses integration tooling to generate a signed API request 
+1. When the API receives a request, it loads the credentials for the calling user from the database and validates the request signature to ensure that the request was not modified in transit
+1. The request is then validated to ensure that a) the request data is correct b) the request passes all validation checks, and c) the user has access to make the change
+1. Assuming the request is in good order, the request is put on a message queue for handling
+1. Engine processes requests on the message queue, sending DNS requests to DNS servers and storing the responses in the database
+
+## Integration
+
+Integrating with VinylDNS is simple since each API endpoint is effectively a distinct DNS operation (eg. create record, update record, delete record, etc.). The only requirement for sending a request is generating the correct AWS SIG4 signature without content length and providing the corresponding HTTP headers so that VinylDNS can verify it. See https://www.vinyldns.io/api/auth-mechanism.html for more details.
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: Instead of exposing the URL can you say "See the API Documentation for more details" with the API Documentation text linked.

File: SYSTEM_DESIGN.md [hunk 19]
```diff
@@ -0,0 +1,52 @@
+# System Design
+
+## Table of Contents
+- [Components](#components)
+- [Process Flow](#process-flow)
+- [Integration](#integration)
+
+## Components
+
+The following diagram illustrates the major components in the VinylDNS ecosystem and the external systems they interact with each other.
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: *that* not *they*

Comment (britneywright) [reply SYSTEM_DESIGN.md]: or take out *each other*

File: SYSTEM_DESIGN.md [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+
+## Process Flow
+
+1. LDAP service authenticates user credentials and grants access to the portal.
+1. If the user is accessing the portal for the first time, user credentials are generated and stored.
+1. User navigates portal or uses integration tooling to generate a signed API request.
+1. When the API receives a request, it loads the credentials for the calling user from the database and validates the request signature to ensure that the request was not modified in transit.
+1. The request is then validated to ensure that a) the request data is correct b) the request passes all validation checks, and c) the user has access to make the change.
+1. Assuming the request is in good order, the request is put on a message queue for handling.
+1. Engine processes requests on the message queue, sending DNS requests to DNS servers and storing the responses in the database.
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: This sentence seems more choppy than the rest of the steps in the flow.

File: SYSTEM_DESIGN.md [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+
+## Integration
+
+Integrating with VinylDNS is simple since each API endpoint is effectively a distinct DNS operation (eg. create record, update record, delete record, etc.). The only requirement for sending a request is generating the correct AWS SIG4 signature without content length and providing the corresponding HTTP headers so that VinylDNS can verify it. See https://www.vinyldns.io/api/auth-mechanism.html for more details.
+
+Current tooling that is available to perform VinylDNS API requests include:
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: "The current tooling available to perform VinylDNS API requests include:"

File: SYSTEM_DESIGN.md [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+1. When the API receives a request, it loads the credentials for the calling user from the database and validates the request signature to ensure that the request was not modified in transit.
+1. The request is then validated to ensure that a) the request data is correct b) the request passes all validation checks, and c) the user has access to make the change.
+1. Assuming the request is in good order, the request is put on a message queue for handling.
+1. Engine processes DNS requests on the message queue, storing the responses in the database.
+
+## Integration
+
+Integrating with VinylDNS is simple since each API endpoint is effectively a distinct DNS operation (eg. create record, update record, delete record, etc.). The only requirement for sending a request is generating the correct AWS SIG4 signature without content length and providing the corresponding HTTP headers so that VinylDNS can verify it. See https://www.vinyldns.io/api/auth-mechanism.html for more details.
+
+The current tooling available to perform VinylDNS API requests include::
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: you have an extra colon here now

File: SYSTEM_DESIGN.md [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+* LDAP Service - application protocol used to authenticate user access to the VinylDNS portal
+
+* Portal - graphical user interface to interact with the VinylDNS API
+
+* Tooling - external libraries and utilities used to interact with the VinylDNS API
+
+## Process Flow
+
+1. LDAP service authenticates user credentials and grants access to the portal.
+1. If the user is accessing the portal for the first time, user credentials are generated and stored.
```
Comment (rebstar6) [inline SYSTEM_DESIGN.md]: say VinylDNS credentials here - 2 types of "user credentials" are mentioned in steps 1 and 2

File: SYSTEM_DESIGN.md [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+1. User navigates portal or uses integration tooling to generate a signed API request.
+1. When the API receives a request, it loads the credentials for the calling user from the database and validates the request signature to ensure that the request was not modified in transit.
+1. The request is then validated to ensure that a) the request data is correct b) the request passes all validation checks, and c) the user has access to make the change.
```
Comment (rebstar6) [inline SYSTEM_DESIGN.md]: make the sub bullets actual bullets

File: SYSTEM_DESIGN.md [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+
+![VinylDNS Architecture Diagram](img/VinylDNS_overview.png)
+
+* API - RESTful endpoints to allow interaction with VinylDNS.
+
+* Database - stores information that the VinylDNS application needs
+
+* DNS servers - communicates DNS changes and resolves DNS records
+
+* Message queue - stores DNS requests
```
Comment (rebstar6) [inline SYSTEM_DESIGN.md]: "stores" sounds like it's a database

the queue here is queuing DNS requests for processing. or "temporarily stores DNS requests for processing"

File: SYSTEM_DESIGN.md [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+1. If the user is accessing the portal for the first time, user credentials are generated and stored.
+1. User navigates portal or uses integration tooling to generate a signed API request.
+1. When the API receives a request, it loads the credentials for the calling user from the database and validates the request signature to ensure that the request was not modified in transit.
+1. The request is then validated to ensure that a) the request data is correct b) the request passes all validation checks, and c) the user has access to make the change.
+1. Assuming the request is in good order, the request is put on a message queue for handling.
+1. Engine processes DNS requests on the message queue, storing the responses in the database.
+
+## Integration
+
+Integrating with VinylDNS is simple since each API endpoint is effectively a distinct DNS operation (eg. create record, update record, delete record, etc.). The only requirement for sending a request is generating the correct AWS SIG4 signature without content length and providing the corresponding HTTP headers so that VinylDNS can verify it. See https://www.vinyldns.io/api/auth-mechanism.html for more details.
```
Comment (rebstar6) [inline SYSTEM_DESIGN.md]: `See [API Authentication](https://www.vinyldns.io/api/auth-mechanism.html) for more details`

File: SYSTEM_DESIGN.md [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+
+The current tooling available to perform VinylDNS API requests include:
```
Comment (rebstar6) [inline SYSTEM_DESIGN.md]: nitpikc: the arrow from Tooling to the API in the diagram is the wrong color

File: SYSTEM_DESIGN.md [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+- [Process Flow](#process-flow)
+- [Integration](#integration)
+
+## Components
+
+The following diagram illustrates the major components in the VinylDNS ecosystem and the external systems they interact with.
+
+![VinylDNS Architecture Diagram](img/VinylDNS_overview.png)
+
+* API - RESTful endpoints to allow interaction with VinylDNS.
```
Comment (britneywright) [inline SYSTEM_DESIGN.md]: should still remove this period if the rest of the bullets don't have them.
